src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt

```diff
@@ -73,53 +83,131 @@ class VisitService(
-            OffenderVisitBalanceAdjustment(
-              offenderBooking = offenderBooking,
-              adjustDate = visitDto.issueDate,
-              adjustReasonCode = VisitOrderAdjustmentReason.VISIT_ORDER_ISSUE,
-              remainingVisitOrders = -1,
-              previousRemainingVisitOrders = offenderVisitBalance.remainingVisitOrders,
-              commentText = "Created by PVB3 for an on-line visit booking",
-              //  authorisedStaffId = visitDto.staffId,
-              //  endorsedStaffId = visitDto.staffId,
-              // TODO: If necessary, could follow stored proc logic and use staffId from:
-              //   select staff_id from staff_user_accounts where username = 'OMS_OWNER';
-            )
+        )
+        visit.visitOrder = VisitOrder(
+          offenderBooking = offenderBooking,
+          visitOrderNumber = visitOrderRepository.getVisitOrderNumber(),
+          visitOrderType = visitOrderTypeRepository.findById(VisitOrderType.pk("PVO")).orElseThrow(),
+          status = visitStatusRepository.findById(VisitStatus.pk("SCH")).orElseThrow(),
+          issueDate = visitDto.issueDate,
+          expiryDate = visitDto.issueDate.plusDays(28),
+        )
+      } else if (offenderVisitBalance.get().remainingVisitOrders!! > 0) {
+        val adjustReasonCode =
+          visitOrderAdjustmentReasonRepository.findById(VisitOrderAdjustmentReason.VO_ISSUE).orElseThrow()
+        offenderVisitBalanceAdjustmentRepository.save(
+          OffenderVisitBalanceAdjustment(
+            offenderBooking = offenderBooking,
+            adjustDate = visitDto.issueDate,
+            adjustReasonCode = adjustReasonCode,
+            remainingVisitOrders = -1,
+            previousRemainingVisitOrders = offenderVisitBalance.get().remainingVisitOrders,
+            commentText = "Created by VSIP for an on-line visit booking",
+            //  authorisedStaffId = visitDto.staffId,
+            //  endorsedStaffId = visitDto.staffId,
+            // TODO: If necessary, could follow stored proc logic and use staffId from:
+            //   select staff_id from staff_user_accounts where username = 'OMS_OWNER';
           )
-        }
+        )
+        visit.visitOrder = VisitOrder(
+          offenderBooking = offenderBooking,
+          visitOrderNumber = visitOrderRepository.getVisitOrderNumber(),
+          visitOrderType = visitOrderTypeRepository.findById(VisitOrderType.pk("VO")).orElseThrow(),
+          status = visitStatusRepository.findById(VisitStatus.pk("SCH")).orElseThrow(),
+          issueDate = visitDto.issueDate,
+          expiryDate = visitDto.issueDate.plusDays(28),
+        )
       }
     }
   }
 
+  private fun cancelBalance(
+    visit: Visit,
+    today: LocalDate,
+  ) {
+    val visitOrder = visit.visitOrder!!
+    val offenderBooking = visit.offenderBooking!!
+    val offenderVisitBalance = offenderVisitBalanceRepository.findById(offenderBooking.bookingId).orElseThrow()
```

Comment:
at some point we should go back and work out what really can be nullable, I guess offenderBooking can never be null

Reply:
Actually it is NOT NULL, so Ill change the mapping